interpreter/exec/int.ml

```diff
@@ -237,7 +237,26 @@ struct
     | 'A' .. 'F' as c ->  0xa + Char.code c - Char.code 'A'
     | _ ->  failwith "of_string"
 
-  let max_upper, max_lower = divrem_u Rep.minus_one ten
+  let max_upper, max_lower =
+    let mup, mlow = divrem_u Rep.minus_one ten in
+    mup, mlow
+
+  let maybe_sign_extend i =
+    (* This module is used with I32 and I64, but the bitwidth can be less
+     * than that, e.g. for I16. When used for smaller integers, the stored value
+     * needs to be signed extened, e.g. parsing -1 into a I16 (backed by Int32)
+     * shoud have all high bits set. We can do that by logor with a mask,
+     * where the mask is minus_one left shifted by bitwidth. But if bitwidth
+     * matches the number of bits of Rep, the shift will be incorrect.
+     *   -1 (Int32) << 32 = -1
+     * Then the logor will be also wrong. So we check and bail out early.
+     * *)
+    if Rep.bitwidth >= 32 then i else
```

Comment:
I'd make this check a little more generic, e.g., by checking whether max_int + 1 = min_int.